url9ecf1e8297_nhamidi_ninja_tgz-pJ8-net_fec_openrq_test_TestRunnerJ8/src/RQLibrary/Encoder.java

```diff
@@ -0,0 +1,2552 @@
+	if ( num_repair_symbols < num_missing_symbols ) {
+	    throw new RuntimeException(Integer.toString(num_missing_symbols - num_repair_symbols)); // TODO shouldn't be runtime exception (too generic)
+	    // throw new RuntimeException( Integer.toString(num_missing_symbols)); // TODO shouldn't be runtime exception (too generic)
+	}
+	
+	// number of extra repair symbols to be used for the decoding process
+	int overhead = num_repair_symbols - missing_symbols.size();
+	
+	// number of rows in the decoding matrix
+	int M = L + overhead;
+	
+	// maps the ESI of a symbol to its line in the decoding matrix
+	Map<Integer, byte[]> esiToLTCode = new TreeMap<Integer, byte[]>();
+	
+	// allocate buffer for the decoded data
+	byte[] decoded_data = new byte[K * T];
+	
+	// all source symbols received! :D
+	if ( num_source_symbols == K ) {
+	    // collect all payloads from the source symbols
+	    for (EncodingSymbol enc_symbol : source_symbols) {
+		System.arraycopy(enc_symbol.getData(), 0, decoded_data, enc_symbol.getESI() * T, T);
+	    }
+	    
+	    // return the "decoded" source block
+	    return new SourceBlock(eb.getSBN(), decoded_data, T, K);
+	}
+	
+	// not so lucky...
+	else {
+	    
+	    // allocate memory for the decoding matrix
+	    byte[][] constraint_matrix = new byte[M][];
+	    
+	    // generate the original constraint matrix
+	    byte[][] lConstraint = generateConstraintMatrix(kLinha, T);
+	    
+	    // copy to our decoding matrix
+	    for (int row = 0; row < L; row++)
+		constraint_matrix[row] = lConstraint[row];
+	    
+	    // initialize D
+	    byte[][] D = new byte[M][T];
+	    
+	    // populate D with the received source symbols
+	    if ( num_source_symbols != 0 ) {
+		
+		Iterator<EncodingSymbol> it = source_symbols.iterator();
+		
+		do {
+		    EncodingSymbol source_symbol = (EncodingSymbol) it.next();
+		    
+		    D[source_symbol.getESI() + S + H] = source_symbol.getData();
+		    
+		} while (it.hasNext());
+	    }
+	    
+	    /*
+	     * for every repair symbol received - replace a missing source symbol's decoding matrix line for its corresponding line - populate D accordingly
+	     */
+	    
+	    Iterator<EncodingSymbol> repair_symbol = repair_symbols.iterator();
+	    
+	    // identify missing source symbols and replace their lines with "repair lines"
+	    for (Integer missing_ESI : missing_symbols) {
+		
+		EncodingSymbol repair = (EncodingSymbol) repair_symbol.next();
+		int row = S + H + missing_ESI;
+		
+		// replace line S + H + missing_ESI with the line for encIndexes
+		Set<Integer> indexes = encIndexes(kLinha, new Tuple(kLinha, repair.getISI(K)));
+		
+		byte[] newLine = new byte[L];
+		
+		for (Integer col : indexes)
+		    newLine[col] = 1;
+		
+		esiToLTCode.put(missing_ESI, constraint_matrix[row]);
+		constraint_matrix[row] = newLine;
+		
+		// fill in missing source symbols in D with the repair symbols
+		D[row] = repair.getData();
+	    }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 536:554, 725:742.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.